mysqld_exporter.go

```diff
@@ -0,0 +1,179 @@
+	duration                   prometheus.Gauge
+	totalScrapes, errorScrapes prometheus.Counter
+	metrics                    map[string]prometheus.Gauge
+}
+
+// return new empty exporter
+func NewMySQLExporter(dsn string) *Exporter {
+	return &Exporter{
+		dsn: dsn,
+		duration: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "exporter_last_scrape_duration_nanoseconds",
+			Help:      "Was the last scrape  duration.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_scrapes_total",
+			Help:      "Current total mysqld scrapes.",
+		}),
+		errorScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "exporter_scrape_errors_total",
+			Help:      "Error mysqld scrapes.",
+		}),
+		metrics: map[string]prometheus.Gauge{},
+	}
+}
+
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	for _, m := range e.metrics {
+		m.Describe(ch)
+	}
+
+	ch <- e.duration.Desc()
+	ch <- e.totalScrapes.Desc()
+	ch <- e.errorScrapes.Desc()
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	scrapes := make(chan []string)
+
+	go e.scrape(scrapes)
+
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.setMetrics(scrapes)
+	ch <- e.duration
+	ch <- e.totalScrapes
+	ch <- e.errorScrapes
+	e.collectMetrics(ch)
+}
+
+func (e *Exporter) scrape(scrapes chan<- []string) {
+	defer close(scrapes)
+
+	now := time.Now().UnixNano()
+
+	e.totalScrapes.Inc()
+
+	db, err := sql.Open("mysql", e.dsn)
+	if err != nil {
+		log.Printf("error open connection to db")
+		e.errorScrapes.Inc()
+		e.duration.Set(float64(time.Now().UnixNano() - now))
+		return
+	}
+	defer db.Close()
+
+	rows, err := db.Query("SHOW STATUS")
+	if err != nil {
+		log.Printf("error running query on db")
+		e.errorScrapes.Inc()
+		e.duration.Set(float64(time.Now().UnixNano() - now))
+		return
+	}
+	defer rows.Close()
+
+	var key, val []byte
+	for rows.Next() {
+		// get RawBytes from data
+		err = rows.Scan(&key, &val)
+		if err != nil {
+			log.Printf("error getting result set")
+			return
+		}
+
+		var res []string = make([]string, 2)
+		res[0] = string(key)
+		res[1] = string(val)
+
+		scrapes <- res
+	}
+
+	e.duration.Set(float64(time.Now().UnixNano() - now))
+
+	return
+}
+
+func (e *Exporter) setMetrics(scrapes <-chan []string) {
+	for row := range scrapes {
+
+		name := strings.ToLower(row[0])
+		value, err := strconv.ParseInt(row[1], 10, 64)
+		if err != nil {
+			// convert/serve text values here ?
+			continue
+		}
+
+		if _, ok := e.metrics[name]; !ok {
+			e.metrics[name] = prometheus.NewGauge(prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      name,
+			})
+		}
+
+		e.metrics[name].Set(float64(value))
+	}
+}
+
+func (e *Exporter) collectMetrics(metrics chan<- prometheus.Metric) {
+	for _, m := range e.metrics {
+		m.Collect(metrics)
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	dsn := os.Getenv("DSN")
```

Comment:
ok
if i need fetch some metrics from sql package or make some special queries for monitoring database application?

in oracle it will be absolutely required.